File: LeetCode/1382. Balance a Binary Search Tree/1382.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* balanceBST(TreeNode* root) {
        vector<int> list;
        
        // Perform inorder traversal to get sorted elements
        inorder(root, list);
        
        // Build balanced BST from sorted elements
        return buildBalancedBST(list, 0, list.size() - 1);
    }

private:
    void inorder(TreeNode* root, vector<int>& list) {
        if (root == nullptr) return;
        
        inorder(root->left, list);  // Traverse left subtree
        list.push_back(root->val);  // Add root value
        inorder(root->right, list); // Traverse right subtree
    }

    TreeNode* buildBalancedBST(const vector<int>& list, int start, int end) {
        if (start > end) return nullptr;
        
        int mid = start + (end - start) / 2;
        TreeNode* root = new TreeNode(list[mid]);
        
        root->left = buildBalancedBST(list, start, mid - 1); // Build left subtree
        root->right = buildBalancedBST(list, mid + 1, end);  // Build right subtree
        
        return root;
    }
};

//other people's solution
/*
class Solution {
    void inorder(TreeNode* root , vector<TreeNode*> &nodes){
        if(root==NULL) return ;
        inorder(root->left,nodes);
        nodes.push_back(root);
        inorder(root->right,nodes);
    }
    TreeNode* solve(vector<TreeNode*> &nodes,int start , int end){
        if(end<start) return NULL;
        int mid = (end-start)/2+start;
        TreeNode* root = nodes[mid];
        root->left = solve(nodes,start,mid-1);
        root->right = solve(nodes,mid+1,end);
        return root;
    }

public:
    TreeNode* balanceBST(TreeNode* root) {
        if(root==NULL) return NULL;
        vector<TreeNode*> inorder_traversal;
        inorder(root,inorder_traversal);
        return solve(inorder_traversal,0,inorder_traversal.size()-1);
    }
};
*/
/*
class Solution {
public:
    TreeNode* balanceBST(TreeNode* root) {
        if (!root) return nullptr;

        // Step 1: Create the backbone (vine)
        // Temporary dummy node
        TreeNode* vineHead = new TreeNode(0);
        vineHead->right = root;
        TreeNode* current = vineHead;
        while (current->right) {
            if (current->right->left) {
                rightRotate(current, current->right);
            } else {
                current = current->right;
            }
        }

        // Step 2: Count the nodes
        int nodeCount = 0;
        current = vineHead->right;
        while (current) {
            ++nodeCount;
            current = current->right;
        }

        // Step 3: Create a balanced BST
        int m = pow(2, floor(log2(nodeCount + 1))) - 1;
        makeRotations(vineHead, nodeCount - m);
        while (m > 1) {
            m /= 2;
            makeRotations(vineHead, m);
        }

        TreeNode* balancedRoot = vineHead->right;
        // Delete the temporary dummy node
        delete vineHead;
        return balancedRoot;
    }

    // Function to perform a right rotation
    void rightRotate(TreeNode* parent, TreeNode* node) {
        TreeNode* tmp = node->left;
        node->left = tmp->right;
        tmp->right = node;
        parent->right = tmp;
    }

    // Function to perform a left rotation
    void leftRotate(TreeNode* parent, TreeNode* node) {
        TreeNode* tmp = node->right;
        node->right = tmp->left;
        tmp->left = node;
        parent->right = tmp;
    }

    // Function to perform a series of left rotations to balance the vine
    void makeRotations(TreeNode* vineHead, int count) {
        TreeNode* current = vineHead;
        for (int i = 0; i < count; ++i) {
            TreeNode* tmp = current->right;
            leftRotate(current, tmp);
            current = current->right;
        }
    }
};
*/

File: LeetCode/1382. Balance a Binary Search Tree/1382.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode BalanceBST(TreeNode root) {
        List<int> list = new List<int>();
        
        // Perform inorder traversal to get sorted elements
        Inorder(root, list);
        
        // Build balanced BST from sorted elements
        return BuildBalancedBST(list, 0, list.Count - 1);
    }

    private void Inorder(TreeNode root, List<int> list) {
        if (root == null) return;
        
        Inorder(root.left, list);  // Traverse left subtree
        list.Add(root.val);        // Add root value
        Inorder(root.right, list); // Traverse right subtree
    }

    private TreeNode BuildBalancedBST(List<int> list, int start, int end) {
        if (start > end) return null;
        
        int mid = start + (end - start) / 2;
        TreeNode root = new TreeNode(list[mid]);
        
        root.left = BuildBalancedBST(list, start, mid - 1); // Build left subtree
        root.right = BuildBalancedBST(list, mid + 1, end);  // Build right subtree
        
        return root;
    }
}

/*
public class Solution {
    List<TreeNode> list = new List<TreeNode>();
    public TreeNode BalanceBST(TreeNode root) {
        InOrder(root);
        if (list.Count <= 2)
            return root;
        return BuildTree(0, list.Count-1);
        
    }

    private TreeNode BuildTree(int l, int r) {
        if(l>r)
            return null;
        
        var m = l+(r-l)/2;
        var root = list[m];

        root.left = BuildTree(l, m-1);
        root.right = BuildTree(m+1, r);
        return root;
    }

    private void InOrder(TreeNode root) {
        if(root == null)
            return;
        InOrder(root.left);
        list.Add(root);
        InOrder(root.right);
    }


}
*/
/*
public class Solution {
    List<int> nodes = new List<int>();
    public TreeNode BalanceBST(TreeNode root) {
        PopulateNodes(root);
        return CreateTree(nodes, 0, nodes.Count-1);
    }

    private void PopulateNodes(TreeNode root){
        if(root==null) return;
        PopulateNodes(root.left);
        nodes.Add(root.val);
        PopulateNodes(root.right);
    }

    private TreeNode CreateTree(List<int> nodes, int start, int end){
        if(start>end) return null;
        int mid = (start+end)/2;
        TreeNode root = new TreeNode(nodes[mid]);
        root.left = CreateTree(nodes, start, mid-1);
        root.right = CreateTree(nodes, mid+1, end);
        return root;
    }
}
*/